src/hera/workflows/script.py

```diff
@@ -654,13 +666,16 @@ def __call__(  # type: ignore [overload-overlap]
         with_param: Optional[Any] = ...,
```

Comment:
* `ArgumentsT` is already `Optional`: https://github.com/argoproj-labs/hera/blob/a50ba238ec3456c56dc6d15a0e0e46b712657108/src/hera/workflows/_mixins.py#L133-L138
* I don't believe `name` is `Optional`; if you pass in `None`, the default will be overwritten: https://github.com/argoproj-labs/hera/blob/a50ba238ec3456c56dc6d15a0e0e46b712657108/src/hera/workflows/script.py#L757-L763

Reply:
> ArgumentsT is already Optional

🤦Oops sorry! 😂 

For the name, your linked code is for the `@script` part, i.e. users can do `@script(name=None)` which is allowed but would be a bit weird - it's allowed for the very niche use case of [inline templates](https://argo-workflows.readthedocs.io/en/latest/inline-templates/) (they are kind of broken/unreliable and we tend to avoid using them though).

The Step/Task name comes from here: https://github.com/argoproj-labs/hera/blob/a50ba238ec3456c56dc6d15a0e0e46b712657108/src/hera/workflows/_meta_mixins.py#L331-L332

But names for Steps/Tasks _are_ mandatory, so my point is if you have a callable function doing `my_function(arguments={...})`, which is allowed (as we infer the name from the `_meta_mixins.py` lines above), wouldn't/shouldn't that raise a linting error?


For completeness, if you do `my_function(name=None,...)` you'd get:
```
examples/workflows/steps/callable_steps.py:14: in <module>
    hello(name=None, arguments={"name": "hello-1-{{inputs.parameters.my-step-input}}"})
src/hera/workflows/script.py:771: in task_wrapper
    return s.__call__(*args, **kwargs)
src/hera/workflows/_meta_mixins.py:378: in __call__
    return Step(template=self, **kwargs)

...

>           raise validation_error
E           pydantic.v1.error_wrappers.ValidationError: 1 validation error for Step
E           name
E             none is not an allowed value (type=type_error.none.not_allowed)
```

Reply:
Whoops! Yep, your code snippet was what I was trying to link to. Let me restate my understanding to check we're on the same page Right now, at runtime:

1. If name is provided and is a string, all is good. 
2. If name is provided and None, there's a runtime error. 
3. If name is not provided, it defaults to the name of the Script. 

I expressed that as `name: str = ...`. i.e. it must be a string but can be left at the default. Declaring it as Optional would only let the user pass in None, which doesn't match current behaviour.

 We could change the implementation to allow None and *then* mark it as Optional in the overloads (I avoided any implementation changes), just for consistency. Is that the ask?

Reply:
> if you have a callable function doing my_function(arguments={...}), which is allowed (as we infer the name from the _meta_mixins.py lines above), wouldn't/shouldn't that raise a linting error?

Not with how I've written the type hints right now, no.

Reply:
Just to further check we're on the same page, `Optional` doesn't let you leave off an argument, it lets you pass in `None`. `= ...` lets you leave off the argument (and has no effect on whether you can pass in `None`).

Reply:
I actually didn't know about the `...` protocol behaviour for defaults, but it seems like the best way to describe "there is a default, but it will be set at runtime (so we can't give an accurate type hint)." (Unless we can capture that? Not fussed right now though.)

I think the code is fine then, as I was just confused by the intellisense suggesting name is required because I didn't understand what the `= ...` type hint actually meant :shipit: